File: implementations/ruby/lib/rcl/parser.rb
require "set"
module RCL
  class Parser
    def initialize(text) = (@text = text; @i = 0)
    def parse
      { "kind" => "document", "blocks" => parse_blocks }
    end
    private
    def parse_blocks(stop = nil)
      out = []
      loop do
        skip_space
        break if eof?
        if stop && peek_identifier == stop
          read_identifier
          break
        end
        out << parse_block
      end
      out
    end

    def parse_block
      name = read_identifier
      arg = parse_optional_string
      expect_identifier("do")
      parse_block_body(name, arg)
    end

    def parse_block_body(name, arg)
      props, blocks, named, seen = {}, {}, [], Set.new
      loop do
        skip_space
        raise "missing end" if eof?
        break if peek_identifier == "end"
        key = read_identifier
        skip_space
        if current == '"' || peek_identifier == "do"
          child_arg = parse_optional_string
          expect_identifier("do")
          child = parse_block_body(key, child_arg)
          child_arg ? named << child : blocks[key] = child
        else
          key = read_dotted_key(key)
          ensure_key_valid!(key, seen)
          expect("=")
          props[key] = parse_value
        end
      end
      expect_identifier("end")
      h = { "kind" => "block", "name" => name, "properties" => props, "blocks" => blocks }
      h["argument"] = arg if arg
      h["named_blocks"] = named unless named.empty?
      h
    end

    def parse_value
      skip_space
      return { "kind" => "string", "value" => read_string } if current == '"'
      return parse_array if current == "["
      return parse_number if number_start?
      id = peek_identifier
      if id == "true" || id == "false"
        read_identifier
        return { "kind" => "boolean", "value" => id == "true" }
      end
      raise "invalid bare value"
    end

    def parse_number
      tok = read_number_token
      { "kind" => "number", "value" => tok.include?(".") ? tok.to_f : tok.to_i }
    end

    def parse_array
      expect("[")
      elems = []
      loop do
        skip_space
        break if current == "]"
        elems << parse_value
        skip_space
        break if current == "]"
        expect(",")
        skip_space
        raise "trailing comma in array" if current == "]"
      end
      expect("]")
      { "kind" => "array", "elements" => elems }
    end

    def parse_optional_string
      skip_space
      current == '"' ? read_string : nil
    end

    def read_dotted_key(key)
      loop do
        skip_space
        break unless current == "."
        advance
        key += ".#{read_identifier}"
      end
      key
    end

    def ensure_key_valid!(key, seen)
      raise "duplicate key #{key}" if seen.include?(key)
      parts = key.split(".")
      seen.each do |existing|
        ex = existing.split(".")
        raise "key conflict #{key} vs #{existing}" if prefix?(parts, ex) || prefix?(ex, parts)
      end
      seen << key
    end

    def prefix?(left, right)
      return false if left.length >= right.length
      left.each_with_index { |p, i| return false if p != right[i] }
      true
    end

    def read_string
      expect('"')
      out = +""
      until eof? || current == '"'
        if current == "\\"
          advance
          raise "unterminated escape" if eof?
          esc = { '"' => '"', "\\" => "\\", "n" => "\n", "t" => "\t" }[current]
          raise "invalid escape" unless esc
          out << esc
        else
          out << current
        end
        advance
      end
      raise "unterminated string" if eof?
      expect('"')
      out
    end

    def read_identifier
      skip_space
      raise "expected identifier" unless current&.match?(/[A-Za-z_]/)
      out = +""
      while !eof? && current.match?(/[A-Za-z0-9_]/)
        out << current
        advance
      end
      out
    end

    def read_number_token
      skip_space
      out = +""
      if current == "-"
        out << "-"
        advance
      end
      raise "invalid number" unless current&.match?(/[0-9]/)
      out << current && advance while !eof? && current.match?(/[0-9]/)
      if current == "." && @text[@i + 1]&.match?(/[0-9]/)
        out << "."
        advance
        out << current && advance while !eof? && current.match?(/[0-9]/)
      end
      out
    end

    def number_start?
      current&.match?(/[0-9]/) || (current == "-" && @text[@i + 1]&.match?(/[0-9]/))
    end

    def peek_identifier
      j = @i
      j += 1 while j < @text.length && @text[j].match?(/\s/)
      return "" unless @text[j]&.match?(/[A-Za-z_]/)
      out = +""
      out << @text[j] && j += 1 while j < @text.length && @text[j].match?(/[A-Za-z0-9_]/)
      out
    end

    def skip_space
      while !eof?
        if current.match?(/\s/) then advance
        elsif current == "#" then advance until eof? || current == "\n"
        else break
        end
      end
    end

    def expect(c) = (skip_space; raise("expected #{c}") unless current == c; advance)
    def expect_identifier(w) = (g = read_identifier; raise("expected #{w}, got #{g}") unless g == w)
    def current = @text[@i]
    def advance = (@i += 1)
    def eof? = (@i >= @text.length)
  end
end

File: implementations/swift/Sources/RCLParser/Parser.swift
public final class Parser {
    private let lexer: Lexer
    private var current: Token

    public init(_ text: String) throws {
        self.lexer = Lexer(text)
        self.current = try lexer.nextToken()
    }

    public func parse() throws -> DocumentNode {
        if current.type == .do {
            try eat(.do)
            let rootValue = try parseArray()
            if current.type != .eof { throw err("unexpected token after root array") }
            return DocumentNode(blocks: [], rootValue: rootValue)
        }
        var blocks: [BlockNode] = []
        while current.type != .eof { blocks.append(try parseBlock()) }
        return DocumentNode(blocks: blocks)
    }

    private func parseBlock() throws -> BlockNode {
        let name = current.value
        try eat(.identifier)
        var argument: String?
        if current.type == .string { argument = current.value; try eat(.string) }
        try eat(.do)

        let out = try parseBlockBody(name: name, argument: argument)
        try eat(.end)
        return out
    }

    private func parseAnonymousBlock() throws -> BlockNode {
        try eat(.do)
        let out = try parseBlockBody(name: "", argument: nil)
        try eat(.end)
        return out
    }

    private func parseBlockBody(name: String, argument: String?) throws -> BlockNode {
        var properties: [String: Node] = [:]
        var blocks: [String: BlockNode] = [:]
        var named: [BlockNode] = []
        var seen: Set<String> = []

        while current.type != .end {
            if current.type == .eof { throw err("missing end") }
            if current.type != .identifier { throw err("expected identifier") }
            let next = try peekToken()
            if next.type == .do {
                let afterDo = try peekToken(offset: 2)
                if afterDo.type == .lbracket {
                    let key = current.value
                    try eat(.identifier)
                    try ensureKeyValid(key, seen: &seen)
                    try eat(.do)
                    properties[key] = try parseArray()
                    try eat(.end)
                } else {
                    let child = try parseBlock()
                    if child.argument != nil { named.append(child) }
                    else { blocks[child.name] = child }
                }
            } else if next.type == .string {
                let child = try parseBlock()
                if child.argument != nil { named.append(child) }
                else { blocks[child.name] = child }
            } else if next.type == .equal || next.type == .dot {
                let key = try parsePropertyKey()
                try ensureKeyValid(key, seen: &seen)
                try eat(.equal)
                properties[key] = try parseValue()
            } else { throw err("invalid statement") }
        }

        return BlockNode(name: name, argument: argument, properties: properties, blocks: blocks, namedBlocks: named)
    }

    private func parsePropertyKey() throws -> String {
        var key = current.value
        try eat(.identifier)
        while current.type == .dot {
            try eat(.dot)
            if current.type != .identifier { throw err("expected identifier after dot") }
            key += ".\(current.value)"
            try eat(.identifier)
        }
        return key
    }

    private func parseValue() throws -> Node {
        switch current.type {
        case .string: let v = current.value; try eat(.string); return .string(StringNode(value: v))
        case .number: let v = Double(current.value) ?? 0; try eat(.number); return .number(NumberNode(value: v))
        case .identifier:
            let v = current.value; try eat(.identifier)
            if v == "true" { return .boolean(BooleanNode(value: true)) }
            if v == "false" { return .boolean(BooleanNode(value: false)) }
            throw err("invalid bare value")
        case .lbracket: return try parseArray()
        case .do: return .block(try parseAnonymousBlock())
        default: throw err("unexpected value")
        }
    }

    private func parseArray() throws -> Node {
        try eat(.lbracket)
        var elements: [Node] = []
        if current.type != .rbracket {
            elements.append(try parseValue())
            while current.type == .comma {
                try eat(.comma)
                if current.type == .rbracket { throw err("trailing comma in array") }
                elements.append(try parseValue())
            }
        }
        try eat(.rbracket)
        return .array(ArrayNode(elements: elements))
    }

    private func eat(_ type: TokenType) throws {
        if current.type != type { throw err("expected \(type), got \(current.type)") }
        current = try lexer.nextToken()
    }

    private func peekToken(offset: Int = 1) throws -> Token {
        let st = lexer.snapshot()
        var tok = current
        for _ in 0..<offset { tok = try lexer.nextToken() }
        lexer.restore(st)
        return tok
    }

    private func err(_ message: String) -> ParseError { ParseError(message, line: current.line, column: current.column) }

    private func ensureKeyValid(_ key: String, seen: inout Set<String>) throws {
        if seen.contains(key) { throw err("duplicate key") }
        let parts = key.split(separator: ".").map(String.init)
        for existing in seen {
            let ex = existing.split(separator: ".").map(String.init)
            if isPrefix(parts, ex) || isPrefix(ex, parts) { throw err("key conflict") }
        }
        seen.insert(key)
    }

    private func isPrefix(_ left: [String], _ right: [String]) -> Bool {
        if left.count >= right.count { return false }
        for i in 0..<left.count where left[i] != right[i] { return false }
        return true
    }
}

public enum RCL {
    public static func parse(_ text: String) throws -> DocumentNode { try RCLCore.parse(text) }
    public static func format(_ text: String) throws -> String { format(try parse(text)) }
    public static func format(_ doc: DocumentNode) -> String { Formatter.format(doc) }
    public static func toObject(_ text: String) throws -> [String: Any] { try toObject(parse(text)) }
    public static func toObject(_ doc: DocumentNode) -> [String: Any] { RCLCore.toObject(doc) }
    public static func toYAML(_ text: String) throws -> String { toYAML(try parse(text)) }
    public static func toYAML(_ doc: DocumentNode) -> String { Converters.toYAML(doc) }
    public static func toTOML(_ text: String) throws -> String { toTOML(try parse(text)) }
    public static func toTOML(_ doc: DocumentNode) -> String { Converters.toTOML(doc) }
    public static func toHCL(_ text: String) throws -> String { toHCL(try parse(text)) }
    public static func toHCL(_ doc: DocumentNode) -> String { Converters.toHCL(doc) }
}
